Allow filtering of files to lint by changes between git branches

pmcnr-hx opened the pull request:
#### What does this PR do? (please provide any background)
Allow filtering of files to lint by changes between git branches. We want to lint our files automatically when we push to GitHub, but we're getting annoyed by how long it takes to lint the whole tree of a big project and just bypass the hook most of the time.

#### What tests does this PR have?
New unit tests in:
* `test/lib/gitBranchFilter.js`
```
gitBranchFilter
  process
    when `git.currentBranch` errors
      ✓ callsback with error
    when `git.nameOnlyDiff` errors
      ✓ callsback with error
    when `git` functions succeed
      ✓ calls `git.nameOnlyDiff` with the current branch
      ✓ adds root level slashes to `git.nameOnlyDiff` result
      ✓ should callback with intersection of files and `git.nameOnlyDiff` result
```
* `test/lib/integrations/eslint.js`
```
_getFilters()
  with null options
    ✓ returns an empty filters list
  with empty options
    ✓ returns an empty filters list
  with a single option
    ✓ returns an array with a single filter function
  with two options
    ✓ returns an array with two filter functions
_filterFilenames()
  with null filters
    ✓ returns unmodified files
  with empty filters
    ✓ returns unmodified files
  with filters
    ✓ applies chain of filters correctly
_processGlobs
  with some files
    with _checkFiles
      ✓ calls since filter if enabled by option
      ✓ calls git since filter if enabled by option
      ✓ calls git branch filter if enabled by option
      ✓ calls all enabled filters
```
* test/lib/util.js
```
addRootLevelSlashes()
  with a root level file
    ✓ adds slashes
  with a nested file
    ✓ does not change it
```
#### How can this be tested?
Checkout this branch and run:
```
./bin/make-up.js -b origin/master bin lib test
```
verify that only the files modified in the PR are linted and compare the differences with plain `make-up`:
```
./bin/make-up.js bin lib test
```

#### Screenshots / Screencast
N/A

#### What gif best describes how you feel about this work?
![Monty Python Flying Mustache Head](http://i.giphy.com/XJKW9YWXyz0MU.gif)

---

- [x] I have checked our general [contributing document](https://github.com/holidayextras/culture/blob/master/CONTRIBUTING.md) and the project specific [contributing document](../blob/master/CONTRIBUTING.md) (if present) and I'm happy for this to be reviewed.

#### Reviewers

**Review 1** (HX Distribution)
- [x] :+1:

**Review 2** (Short Breaks)
- [x] :+1:

**Review 3** _(optional)_
- [ ] :+1:

By adding a +1 you are confirming you have...
- Witnessed the work behaving as expected (this could be on the author's machine or screencast).
- Checked for coding anti-patterns.
- Checked for appropriate test coverage.
- Checked all the tests are passing.

\*  for HX this review must be completed by an SE, SA or Project Guru

hx-markterry commented:
has this been tested when someone else merges into master? or this code is used on a merge commit?

pmcnr-hx commented:
My use case is for the `pre-push` where we will be running it in a branch off usually either `master` or `staging` (when we move to use that with the new CD strategy). Running this before it leaves the local repository saves time for developers because they won't have to wait for CI to run linting and go check the status on GitHub. On CI, because the extra seconds don't really matter much, I was just expecting we would simply lint the whole codebase like we do now. Does that answer your question @hx-markterry?

hx-markterry commented:
Spoken to @pmcnr-hx about this and have agreed on the way forward, so waiting for new commits on this PR.

pmcnr-hx commented:
@hx-markterry: I have made some tweaks to the previous commits and removed the `GitSinceFilter` as discussed. I have added a note to the README, that if the branch (usually `master`) off which the feature branch was created has new commits, the files changed in the new commits will also be linted until it is merged or rebased on the origin branch, which is expected and acceptable. After the merge/rebase it behaves as expected only linting the changed files between branches.

hx-markterry commented on CHANGELOG.md:
```diff
@@ -1,5 +1,5 @@
-* **7.2.0** - Updated eslint to 1.10.3 and eslint-plugin-react to 3.16.1.
-			- Updated rules in configs/scss-lint.yml. If used with SCSS-Lint, you need version 0.34+
+* **8.0.9** - Allow filtering of files to lint by changes between git branches. Remove git since filter.
```
this should be `8.0.0` to match the package.json

pmcnr-hx replied:
Oopsie... Slip of the finger. I meant `8.0.0`!

hx-markterry commented:
Ok with me, just need to resolve the above minor comment :roller_coaster:

philmeehan commented:
@pmcnr-hx :+1: from me - really good work. Thanks for making it an option :)
